[SPARK-32975][K8S] Add config for driver readiness timeout before executors start

cchriswu opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error message, please read the guideline first:
     https://spark.apache.org/error-message-guidelines.html
-->

### What changes were proposed in this pull request?
Add a new config that controls the timeout of waiting for driver pod's readiness before allocating executor pods. This wait only happens once on application start.
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
The driver's headless service can be resolved by DNS only after the driver pod is ready. If the executor tries to connect to the headless service before driver pod is ready, it will hit UnkownHostException and get into error state but will not be restarted. **This case usually happens when the driver pod has sidecar containers but hasn't finished their creation when executors start.** So basically there is a race condition. This issue can be mitigated by tweaking this config.
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce _any_ user-facing change?
A new config `spark.kubernetes.allocation.driver.readinessTimeout` added.
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->


### How was this patch tested?
Exisiting tests.
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

cchriswu committed:
Ensure driver is ready before executors start

Before creating executor pods, wait until driver gets ready. The
driver's headless service can be resolved by DNS only after the driver
pod is ready. If the executor tries to connect to the headless service
before driver pod is ready, it will hit UnkownHostException and get into
error state but will not be restarted.

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
Thank you, @cchriswu !

dongjoon-hyun reviewed (dismissed):
If you put this here, this will be called everything at `onNewSnapshots`. The driver src creation check should be done once. Could you move the check logic into a better place?

cchriswu committed:
Move the wait for driver readiness

Move the wait from onNewSnapshots() to start() in ExecutorPodsAllocator,
so that this check will be only done once on start.

Mock waitUntilReady() to not break existing tests

cchriswu reviewed (commented):
Thank you for the comment, @dongjoon-hyun!
I moved the check inside `start()` so that it will only run once.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:
```diff
@@ -99,6 +102,16 @@ private[spark] class ExecutorPodsAllocator(
   @volatile private var deletedExecutorIds = Set.empty[Long]
 
   def start(applicationId: String, schedulerBackend: KubernetesClusterSchedulerBackend): Unit = {
+    // wait until the driver pod is ready to ensure executors can connect to driver svc
```
Can we be more specific? The problem is the absence of K8s's headless service resource for this driver pod. For example, since K8s is asynchronously working, the problem can happen even when the driver pod is ready with all sidekicks and the K8s service is not ready to work with this driver pod.

cchriswu replied:
Thank you for the review. Makes sense to me. Addressed in the newer commit.

dongjoon-hyun commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocatorSuite.scala:
```diff
@@ -104,6 +105,8 @@ class ExecutorPodsAllocatorSuite extends SparkFunSuite with BeforeAndAfter {
     when(podOperations.withName(driverPodName)).thenReturn(driverPodOperations)
     when(driverPodOperations.get).thenReturn(driverPod)
+    when(driverPodOperations.waitUntilReady(anyInt(), any(classOf[TimeUnit])))
+      .thenReturn(driverPod)
```
Can we simplify like the following?
```
when(driverPodOperations.waitUntilReady(any(), any())).thenReturn(driverPod)
```

cchriswu replied:
Sure. Thanks!

cchriswu committed:
Improve comment and code style

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:
```diff
@@ -61,6 +62,8 @@ private[spark] class ExecutorPodsAllocator(
     podAllocationDelay * 5,
     conf.get(KUBERNETES_ALLOCATION_EXECUTOR_TIMEOUT))
 
+  private val driverPodReadinessTimeout = 5
```
Like KUBERNETES_ALLOCATION_EXECUTOR_TIMEOUT, we need an official configuration instead of the magic number here. Could you add it to `resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala`?

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:
```diff
@@ -99,6 +102,17 @@ private[spark] class ExecutorPodsAllocator(
+      case e: InterruptedException =>
+        logWarning(s"Timeout waiting for driver pod ${kubernetesDriverPodName.get} get ready in " +
+          s"namespace $namespace")
+    }
```
If we are going to just `logWarning`, the following code pattern will be better like the other places.
```scala
Utils.tryLogNonFatalError {
  kubernetesClient.pods()
    .withName(kubernetesDriverPodName.get)
    .waitUntilReady(driverPodReadinessTimeout, TimeUnit.MINUTES)
}
```

cchriswu replied:
If we just do `logWarning` and continue to create executor pods after this timeout, it's very likely executor pods will still hit the `UnknownHostException` issue.
Considering this, throwing a SparkException may be more intuitive and fail-fast. Please let me know your thoughts on this. Thanks.
Like here:
https://github.com/apache/spark/blob/0342dcb6289bae67f6ab742b8fd03b2d653e52ea/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala#L73-L79

dongjoon-hyun replied:
-1 for `SparkException` idea. That is a huge regression.

`UnknownHostException` is not a job blocker because Spark is going to create the required executors in any way eventually. If you throws SparkException, you change will be a breaker of the existing Spark pipelines indeed.

cchriswu replied:
As the [JIRA issue](https://issues.apache.org/jira/browse/SPARK-32975) describes, driver will not create new executors when this error happens.
> Once this error happens, driver doesn't restart executor.

However, other executors may be running fine. It's just that the number of healthy executors will be less than expected.    So I agree not to use `SparkException` here.

dongjoon-hyun replied:
I don't think the error report is correct because Apache Spark doesn't use `restart` executor at all. The driver will keep creating new executors because the executors with `UnknownHostException` will be dead on the failure. I'm currently describing the Apache Spark behavior, not 3rd party operator behavior.

Commit:
Add config for the timeout

dongjoon-hyun commented:
Thank you for updating.

dongjoon-hyun commented:
ok to test

cchriswu committed:
Fix lint issue

SparkQA commented:
**[Test build #139310 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139310/testReport)** for PR 32752 at commit [`9fafec3`](https://github.com/apache/spark/commit/9fafec335e674f87a98f4e8f90dccc6800b17b80).

SparkQA commented:
**[Test build #139312 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139312/testReport)** for PR 32752 at commit [`20509d1`](https://github.com/apache/spark/commit/20509d147102819d1b2a28d02c1c2437ab989668).

dongjoon-hyun commented:
We are almost close now, @cchriswu . Could you update the PR title and description according to the last of this PR code? For example, this is only a best-effort approach and `ensure` looks misleading.

SparkQA commented:
**[Test build #139310 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139310/testReport)** for PR 32752 at commit [`9fafec3`](https://github.com/apache/spark/commit/9fafec335e674f87a98f4e8f90dccc6800b17b80).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139310/

SparkQA commented:
**[Test build #139312 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139312/testReport)** for PR 32752 at commit [`20509d1`](https://github.com/apache/spark/commit/20509d147102819d1b2a28d02c1c2437ab989668).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139312/

cchriswu commented:
Title and description changed. Thank you for your reviews, @dongjoon-hyun

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43834/

dongjoon-hyun commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:
```diff
@@ -300,6 +300,17 @@ private[spark] object Config extends Logging {
+      .doc("Time to wait for driver pod to get ready before creating executor pods. This wait " +
+        "only happens on application start. If timeout happens, executor pods will still be " +
+        "created.")
+      .version("3.2.0")
```
@cchriswu . Sorry, but could you update this to `3.1.3`? Since K8s is GA at Apache Spark 3.1 and this bug is reported at 2.4.x, I believe we need to backport this to branch-3.1 and release `3.1.3` and `3.2.0` together.

cchriswu replied:
Sure . No worries. Updated :)

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43834/

cchriswu committed:
Update config version to 3.1.3

This config needs to be backported to branch-3.1.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43834/

SparkQA commented:
**[Test build #139321 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139321/testReport)** for PR 32752 at commit [`4dbae0a`](https://github.com/apache/spark/commit/4dbae0a7280471df9903c9a4b2a998f915968bc9).

SparkQA commented:
**[Test build #139321 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139321/testReport)** for PR 32752 at commit [`4dbae0a`](https://github.com/apache/spark/commit/4dbae0a7280471df9903c9a4b2a998f915968bc9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139321/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43843/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/43843/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/43843/

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, @cchriswu .
Merged to master/branch-3.1 for Apache Spark 3.2.0 and 3.1.3.

dongjoon-hyun commented:
Congratulation for your first contribution and welcome to the Apache Spark community, @cchriswu . 
What is your Apache JIRA id? I want to add you to the Apache Spark contributor group and assign SPARK-32975 to you.

cchriswu commented:
Thank you @dongjoon-hyun. My Apache JIRA id is cchriswu :)

dongjoon-hyun commented:
Thank you! You are added to the Apache Spark contributor group and SPARK-32975 is assigned to you. :)
